Copy Site: Add back functionality on domains step

kozer opened the pull request:
See p4TIVU-aoH-p2#comment-11701
Related to https://github.com/Automattic/dotcom-forge/issues/1519

Original pr: https://github.com/Automattic/wp-calypso/pull/72327

#### Proposed Changes

Refactor of the `DomainsStep` component to solve the bug where the user needs to click twice to the bug button to go back to the `/sites` page.

The step has been split into two components: the main component which renders the step wrapper (which includes the header), and a child component which renders the domain picker. The component that renders the header no longer needs to subscribe to the `domainForm` value from the onboarding store. Each time the domain search field is unfocused `domainForm` gets resaved, and this is what was causing the header to rerender.

#### Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->
##### Test Copy site flow
1. Make sure you have a site with `Business` plan and atomic.
2. Navigate to `/sites` 
3. Click the dots menu on the right (at your site's row)
4. You should be navigated to the domain selection screen.
5. On the top right there is a `Back` button. Click it **once**, and ensure that you navigate back to the `/sites` page

##### Test Link in bio flow

Because of the refactor, makes sense to take a look at `link-in-bio` flow as well to see that the component works as expected:

1. Navigate to `/setup?flow=link-in-bio`
2. Take the flow until you land to the domain selection screen.
3. Ensure that the input is working (search for a domain and ensure that it works)
4. Test that the `Decide later` button works (should skip this step)
5. Make sure that the filters are working.
6. Select a free domain and ensure that the process completes successfully.
7. Ensure you select a free domain and a plan other than Free. Make sure you can checkout successfully.

#### Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

fixes https://github.com/Automattic/dotcom-forge/issues/1519

kozer committed:
update: Add back functionality on domains step

sejas committed:
Copy Site: use `exitFlow` in domains step back button (#72450)

* Copy Site: use exitFlow to goBack on domains step

* Copy Site: rename domain step flowName

* Copy Site: set default exitflow to /sites

kozer committed:
Update: Restore onClick and add a onMouseDown to prevent onBlur from firing

ciampo commented:
Hey @kozer , us (Team Components) usually work on the `@wordpress/components` package and therefore don't have a lot of familiarity with Calypso and its onboarding/landing areas. Is there anyone else that you could tag to help review this PR?

Thank you!

p-jackson committed:
Unfocusing the domain component shouldn't rerender the whole step

Rebasing the domains step into two components so that the component that
renders the header no longer needs to subscribe to the `domainForm` in
the onboarding store. Each time the domain search field is unfocused the
`domainForm` is updated, which caused the header to rerender.

p-jackson commented:
@kozer The original refactor was a bit messy so I've tidied up a bit by moving all the domain-picking code into the child component. I've also split each component into its own file.

I've also removed the code which tidied up the `onMouseDown` hack. I think that should be done in #72327 rather than this PR. We could actually rebase this PR now so that it's based on `trunk` if you wanted. Up to you.

@Automattic/vertex FYI we're refactoring the domain step to remove problematic re-renders. There shouldn't be any changes in behaviour, just want to give you a heads up and an opportunity to take a look at the change.

kozer commented:
@p-jackson 

> @kozer The original refactor was a bit messy so I've tidied up a bit by moving all the domain-picking code into the child component. I've also split each component into its own file.
> 
> I've also removed the code which tidied up the `onMouseDown` hack. I think that should be done in #72327 rather than this PR. We could actually rebase this PR now so that it's based on `trunk` if you wanted. Up to you.
> 
> @Automattic/vertex FYI we're refactoring the domain step to remove problematic re-renders. There shouldn't be any changes in behaviour, just want to give you a heads-up and an opportunity to take a look at the change.

@p-jackson as this branch contains all the changes I did on my branch and solve the issue, I don't see the value of having my branch open. I'll rebase this pr, update the description, and close mine.

I'll also remove the `onMouseDown` and have this as the main branch that solves the issue.

Thank you very much for your contribution.

kozer committed:
Update: remove onMouseDown leftover

p-jackson reviewed (approved):
Caveat: I did the refactor so take this review with a grain of salt.

I've tested the domain step and I can't find any regressions so I think the refactor is good 🤞

The other flows I found that use this step are `/setup/ecommerce` and `/setup/link-in-bio-tld` and I couldn't find a way to break them either.

I'd hoped to test the case where you already own the domain you want to use, but afaict none of these flows supported domain mapping to begin with.

kozer commented:
> I've tested the domain step and I can't find any regressions so I think the refactor is good crossed_fingers

Me neither. I guess we are good. I hope that on Monday others will check it as well, so we will be more confident.

kozer committed:
Merge branch 'trunk' into add/back_button_domains_page_refactor_domain_step

sejas commented on client/landing/stepper/declarative-flow/internals/steps-repository/domains/index.tsx:
```diff
@@ -291,186 +204,30 @@ const DomainsStep: Step = function DomainsStep( { navigation, flow } ) {
 			isWideLayout={ true }
-			hideBack={ true }
+			shouldStickyNavButtons={ isCopySiteFlow( flow ) }
+			hideBack={ ! isCopySiteFlow( flow ) }
```
We can consider changing the text to "Back to Sites".

```suggestion
			hideBack={ ! isCopySiteFlow( flow ) }
			backLabelText={ __( 'Back to Sites' ) }
```
<img width="1382" alt="Screenshot 2023-02-06 at 13 20 08" src="https://user-images.githubusercontent.com/779993/216982461-f1895777-793f-49e3-b195-a7f59f5d0403.png">

sejas reviewed (approved):
I tested it and it works as expected. I couldn't find any regression due the big refactor. I tried `/setup/link-in-bio` nor `/setup/ecommerce` and those still working as before.

My only comment is that this back button is different from import flow:
https://github.com/Automattic/wp-calypso/pull/72327#issuecomment-1401555490

We could consider changing the text to `Back to Sites`. It will improve the experience when the flow is triggered from `Settings General > Site Tools`.
<img width="742" alt="Screenshot 2023-02-06 at 13 26 56" src="https://user-images.githubusercontent.com/779993/216983342-622e77fd-3458-4130-8635-df4281d7b812.png">

wojtekn commented:
I've rechecked it, and I see two things:

- we use different label and text for "Back" button in different flows

| Standard flow | Copy Site flow |
|----|----|
| <img width="1237" alt="Screenshot 2023-02-06 at 14 26 52" src="https://user-images.githubusercontent.com/727413/216983998-62fed5f8-1bc7-4b90-bec5-71cec85b1f2b.png"> | <img width="1240" alt="Screenshot 2023-02-06 at 14 27 41" src="https://user-images.githubusercontent.com/727413/216984010-a0e19b81-150c-4b36-8fa6-bfa68adddf78.png"> |

- when I load/reload domain step in Copy Site flow, "Copy Site" appears near WP logo for a short time, and then disappears

![Screen Capture on 2023-02-06 at 14-33-09](https://user-images.githubusercontent.com/727413/216984714-c893b51f-6f02-45b7-9d0e-8ecddc3200f2.gif)

kozer committed:
Merge branch 'trunk' into add/back_button_domains_page_refactor_domain_step

kozer committed:
Update `backLabelText` in domains step

Co-authored-by: Antonio Sejas <antonio@sejas.es>

kozer commented:
> we use different label and text for "Back" button in different flows

Fixed.

>when I load/reload domain step in Copy Site flow, "Copy Site" appears near WP logo for a short time, and then disappears

This is how the `StepContainer` works. Let's keep the scope limited, and open another issue for that.

danielbachhuber commented:
> This is how the `StepContainer` works. Let's keep the scope limited, and open another issue for that.

@kozer Can you open this issue?

danielbachhuber commented:
@kozer Also, given this is a somewhat risky refactor, what are the key metrics we should monitor post-deploy?

kozer commented:
> @kozer Also, given this is a somewhat risky refactor, what are the key metrics we should monitor post-deploy?

Taking a look at the code, I found the following events,  that we can check to see if there is a huge discrepancy between the pre and post-deploy.

It's the `ga` event with `category`: `Domain Search`
and the following events on our side: `calypso_domain_search_submit_step`, `'calypso_signup_skip_step'`

kozer committed:
Merge branch 'trunk' into add/back_button_domains_page_refactor_domain_step

danielbachhuber commented:
> Taking a look at the code, I found the following events, that we can check to see if there is a huge discrepancy between the pre and post-deploy.

@kozer Great, thank you. Can you report the results to `#dotcom-yolo` in a few hours, and again tomorrow?

When you do, you can link to the Slack messages from here. It's better to report the data in Slack so we can keep it private.

kozer commented:
@danielbachhuber , today I'm leaving a bit early unfortunately. Sorry about that. Tomorrow I will be on top of that all day

danielbachhuber commented:
@kozer Ok. In the future, please make sure you're around for at least a few hours after landing a major code change.

danielbachhuber commented:
@kozer Another thing I just noticed (that's apparent in @wojtekn's screenshot too): the action bar has a white background, even though the domains background is light gray

![image](https://user-images.githubusercontent.com/36432/217940224-fd90d296-4e62-4686-8583-e4a30a798104.png)

danielbachhuber commented:
> the action bar has a white background, even though the domains background is light gray

Created a follow-up issue for this: https://github.com/Automattic/dotcom-forge/issues/1622

kozer commented:
> @kozer Another thing I just noticed (that's apparent in @wojtekn's screenshot too): the action bar has a white background, even though the domains background is light gray
> 
> ![image](https://user-images.githubusercontent.com/36432/217940224-fd90d296-4e62-4686-8583-e4a30a798104.png)
> 

@danielbachhuber I agree, but this is not relevant with the scope of the pr. This is how it was implemented originally. Do you want me to create a new issue for that?
